fix(ast/estree): fix `BindingIdentifier`

hi-ogawa opened the pull request:
spec: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/expression/Identifier/spec.ts

I'm not sure whether this can be fixed on struct side `BindingIdentifier` instead of field side `Class.id`. `BindingIdentifier` is used inside flatten `BindingPatternKind` enum, so changing it might be complicated.

There are probably many instances like this, so I'm just trying to see the pattern first.

overlookmotel commented:
Personally I would like to rewrite `BindingPattern` entirely to get rid of the flattened enum `BindingPatternKind`. That's the only place we have to flatten an enum in the entire AST, and this oddity complicates our code in various places. Also there's no `Span` for the entire node including the type annotation.

Something like:

```rs
enum BindingPattern {
    BindingIdentifier(Box<'a, BindingIdentifierWithType<'a>>),
    ObjectPattern(Box<'a, ObjectPatternWithType<'a>>),
    ArrayPattern(Box<'a, ArrayPatternWithType<'a>>),
    AssignmentPattern(Box<'a, AssignmentPatternWithType<'a>>),
}

struct BindingIdentifierWithType<'a> {
    pub span: Span,
    pub ident: BindingIdentifier<'a>,
    pub type_annotation: Option<Box<'a, TSTypeAnnotation<'a>>>,
    pub optional: bool,
}

struct ObjectPatternWithType<'a> {
    pub span: Span,
    // `ObjectPattern` is a separate type because it also has a `Span`
    pub pattern: ObjectPattern<'a>,
    pub type_annotation: Option<Box<'a, TSTypeAnnotation<'a>>>,
    pub optional: bool,
}

// Same for `ArrayPatternWithType` and `AssignmentPatternWithType`
```

This has the downside of requiring more AST types. But the advantage is that we could make illegal syntax impossible to represent in the AST. e.g. the type annotation is not legal here:

```ts
[{x}: Foo] = array;
```
[TS playground](https://www.typescriptlang.org/play/?#code/DYUwLgBAhgTjBc0B2BPCBeCBtAugbgChRIAPRKVQrAbxIF9zUcNo5Cg)

But... problem is that `BindingPattern` is so commonly used, it'd require tons of changes (including downstream in Rolldown). So we had decided to leave it.

Anyway... TLDR... if `BindingPattern` has to change to enable TS-ESLint alignment, then doing that would not be out of the question, in my opinion.

hi-ogawa commented:
It looks like `BindingIdentifier` can also do same as `IdentifierName` and `IdentifierReference`. The catch is that, when `BindingIdentifier` is inside `BindingPattern`, raw json string has duplicate keys for `typeAnnotation` and `optional`, but `JSON.parse` picks up the last (which is the correct value from `BindingPattern`), so parse result in js world looks correct.

```ts
let x: string = "foo"
```

```sh
$ node napi/parser/example.mjs test.ts
          "id": {
            "type": "Identifier",
            "start": 45,
            "end": 54,
            "name": "x",
            "decorators": [],
            "optional": false,
            "typeAnnotation": null,
            "typeAnnotation": {
              "type": "TSTypeAnnotation",
              "start": 46,
              "end": 54,
              "typeAnnotation": {
                "type": "TSStringKeyword",
                "start": 48,
                "end": 54
              }
            },
            "optional": false
          },
```

I'm not sure if we should concern with this right now or proceed with this until we see issues with this approach.

overlookmotel reviewed (approved):
The duplicate fields in JSON are not ideal, and we'll need to fix at some point. But let's go with it for now, and fix it later once we have achieved alignment of the whole AST.

If it turns out we have to change the shape of `BindingPattern` anyway, then that'll make it easier to fix.

overlookmotel commented:
### Merge activity

* **Mar 21, 1:44 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

hi-ogawa committed:
fix(ast/estree): fix `BindingIdentifier` (#9822)

spec: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/expression/Identifier/spec.ts

I'm not sure whether this can be fixed on struct side `BindingIdentifier` instead of field side `Class.id`. `BindingIdentifier` is used inside flatten `BindingPatternKind` enum, so changing it might be complicated.

There are probably many instances like this, so I'm just trying to see the pattern first.

ArnaudBarre commented on npm/oxc-types/types.d.ts:
```diff
@@ -70,6 +70,9 @@ export interface IdentifierReference extends Span {
 export interface BindingIdentifier extends Span {
   type: 'Identifier';
   name: string;
+  decorators?: [];
```
Can this types be generated as non optional?
